README

.Oo Sally - A Tool for Embedding Strings in Vector Spaces
    Copyright (C) 2010-2012 Konrad Rieck (konrad@mlsec.org)

Introduction
-- 

  Sally is a small tool for mapping a set of strings to a set of vectors. 
  This mapping is referred to as embedding and allows for applying
  techniques of machine learning and data mining for analysis of string
  data.  Sally can be applied to several types of strings, such as text
  documents, DNA sequences or log files, where it can handle common formats
  such as directories, archives and text files of string data.

  Sally implements a standard technique for mapping strings to a vector
  space that is often referred to as vector space model or bag-of-words
  model.  The strings are characterized by a set of features, where each
  feature is associated with one dimension of the vector space.  The
  following types of features are supported by Sally: bytes, words, n-grams
  of bytes and n-grams of words.

  Sally proceeds by counting the occurrences of the specified features in
  each string and generating a sparse vector of count values. 
  Alternatively, binary or TF-IDF values can be computed and stored in the
  vectors.  Sally then normalizes the vector, for example using the L1 or L2
  norm, and outputs it in a specified format, such as plain text or in
  LibSVM or Matlab format.

  Consult the manual page of Sally for more information.

Dependencies
--

  >= zlib-1.2.1
  >= libconfig-1.4
  >= libarchive-2.70

  Debian and Ubuntu Linux:  libz-dev libconfig8-dev libarchive-dev
  MacOS X with Homebrew:    libconfig libarchive (homebrew-dupes)
  MacOS X with MacPorts:    libconfig-ht libarchive

  The source code of the libraries is available here:
  zlib        http://www.zlib.net/
  libconfig   http://www.hyperrealm.com/libconfig/
  libarchive  http://libarchive.github.com/

Compilation & Installation
--

From GIT repository first run

  $ ./bootstrap

From tarball run

  $ ./configure [options]
  $ make
  $ make check
  $ make install

Options for configure

  --prefix=PATH           Set directory prefix for installation

  By default Sally is installed into /usr/local. If you prefer a
  different location, use this option to select an installation
  directory.

  --enable-libarchive     Enable support for loading archives
     
  If this feature is enabled, Sally can also be applied to read the
  contents of archives, such as .tgz and .zip.  This allows for
  processing string data in compressed form and may drastically save
  storage space.

  --enable-openmp         Enable support for OpenMP (experimental)
    
  This feature enables support for OpenMP in Sally. It is still
  experimental.  Sally will execute certain parts of the processing in
  parallel making use of multi-core architectures where possible.
 
  --enable-md5hash        Enable MD5 as alternative hash

  Sally uses a hash function for mapping different features to
  different dimensions in the vector space.  By default the very
  efficient Murmur hash is used for this task.  In certain critical
  cases it may be useful to use a cryptographic hash as MD5.


CHANGES
2012-12-27  Konrad Rieck <konrad@mlsec.org>

    - updated autotool config

2012-12-18  Konrad Rieck <konrad@mlsec.org>

    - ubuntu switched to libconfig9

2012-09-28  Konrad Rieck <konrad@mlsec.org>

    - Added authors to documentation

2012-08-28  Konrad Rieck <konrad@mlsec.org>

    - added misc docs to manual

2012-08-24  chwress <chwress@github>

    - fix some typos in the code documentation

    - nicely make use of TRUE & FALSE defines

    - make fvec_equals actually work (thx konrad)

2012-08-23  Konrad Rieck <konrad@mlsec.org>

    - new switch for default config is D

2012-08-23  chwress <chwress@github>

    - introduce idf_check for being able to properly unit-test the internal
    computed idf values

    - introduce the fvec_equals(.,.) function

    - the 1-pass of the tf*idf embedding (determining the idf values) must not
    "post-process" the raw feature vector

2012-08-22  chwress <chwress@github>

    - unify the feature vector creation

    - introduce fvec_truncate

2012-08-21  Konrad Rieck <konrad@mlsec.org>

    - implemented thresholding of vectors

    - config for thresholding

    - docs for thresholding

    - minor fixes

2012-08-21  chwress <https://github.com/chwress>

    - update the eMail address of chwress

    - naming ("print_default" -> "print_defaults")

2012-08-20  Konrad Rieck <konrad@mlsec.org>

    - added timing output

2012-08-17  Konrad Rieck <konrad@mlsec.org>

    - re-enabled openmp code

    - exit if stop word file no available but requested

    - refactored delimiter setup

    - basic framework for stopwords

    - added missing header (thanks to Andreas Ziehe)

    - stop words require the definition of delimiters

    - added missing file to automake

    - added run-time eval code to master

    - added documentation for stop words

    - support for configuring stop words

2012-08-16  Konrad Rieck <konrad@mlsec.org>

    - support for reversing of strings

    - fixed wrong email in docs

    - removed global configuration file

    - support for printing the default configuration

2012-08-13  Konrad Rieck <konrad@mlsec.org>

    - added christian as author

    - added new authors

    - fixes to stdin module

2012-08-09  Alexander Bikadorov <abiku@cs.tu-berlin.de>

    - added stdin input module

2012-08-06  Konrad Rieck <konrad@mlsec.org>

    - synchronized sally.cfg and sconfig.c (thx to chwress)

2012-05-26  Konrad Rieck <konrad@mlsec.org>

    - added md README

2012-05-14  Konrad Rieck <konrad@mlsec.org>

    - minor fixes to man page

    - fixed typo

    - fixed configure

2012-05-13  Konrad Rieck <konrad@mlsec.org>

    - fixed broken md5 copy

    - fixed gzfile pointer madness

    - added note about zlib dep

    - new release

    - spell check

2012-04-23  Konrad Rieck <konrad@mlsec.org>

    - code cleanup 6

    - code cleanup 5

    - code cleanup 4

    - code cleanup 3

    - code cleanup 2

    - code cleanup 1

2012-03-01  Konrad Rieck <konrad@mlsec.org>

    - changed legend

    - some results

2012-02-21  Konrad Rieck <konrad@mlsec.org>

    - removed some brackets

    - support for signed embedding added

    - added framework for new option

2012-02-10  Konrad Rieck <konrad@mlsec.org>

    - first cleanup of lib dir

    - destroyed structure. hmm.

2012-02-05  Konrad Rieck <konrad@mlsec.org>

    - code for sorting bytes in n-grams

    - code for sorting words in n-grams

    - added config switches

    - added support for decoding strings with URI encoding

    - switched to libtoolize

2012-02-04  Konrad Rieck <konrad@mlsec.org>

    - moved decode function to util

2012-01-22  Konrad Rieck <konrad@mlsec.org>

    - sync

2011-10-19  Konrad Rieck <konrad@mlsec.org>

    - half version of positional n-grams. explicit reprepresentation at least
    incorrect

    - config for positional n-grams

2011-07-09  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added note about dimensionality

2011-07-05  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - first version of output module for cluto

2011-04-01  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added missing config file

2011-02-21  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - new release approaching

2011-01-17  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - linux fixes

2011-01-06  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added support for global configuration and command line options

2011-01-04  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - changed names of format parameters

    - fixed regex matching

    - fixed regex code

    - added support for labels in text lines

2011-01-02  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - extended example configuration

2010-12-17  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - sanitized contributions directory

    - new contrib

2010-10-08  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - missing latex templates

    - added support to generate PDF manual

2010-10-07  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - new release 0.5.2

2010-10-05  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added contrib directory and simple matlab function

2010-10-01  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - removed trim functionality as it badly interfers with normalization and
    tfidf

    - support to trim min. and max. of dimensions

    - new minor update

    - support for exporting features to matlab

    - added label to matlab struct

    - feature vectors are now converted to matlab structs

    - improved documentation of output and input modules

2010-09-30  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated documentation

2010-09-29  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - release version 0.5.0

2010-09-28  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - another minor fix to manual

    - further extensions to manual

    - minor fix to manual

2010-09-23  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added support to extract labels from FASTA descriptions

src/md5.c
/*
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.  This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 */

#include "config.h"
#include "common.h"
#include "md5.h"

/*
 * Wrapper for MD5 function compatible to OpenSSL interface.
 * 'buf' needs to provide an allocated array of 16 bytes for 
 * the digest.
 */
void MD5(unsigned char *x, unsigned l, unsigned char *buf)
{
    struct MD5Context ctx;

    MD5Init(&ctx);
    MD5Update(&ctx, x, l);
    MD5Final(buf, &ctx);
}

#ifndef HIGHFIRST
#define byteReverse(buf, len)   /* Nothing */
#else
void byteReverse(unsigned char *buf, unsigned uint32_t longs);

#ifndef ASM_MD5
/*
 * Note: this code is harmless on little-endian machines.
 */
void byteReverse(unsigned char *buf, unsigned uint32_t longs)
{
    uint32_t t;
    do {
        t = (uint32_t) ((unsigned) buf[3] << 8 | buf[2]) << 16 |
            ((unsigned) buf[1] << 8 | buf[0]);
        *(uint32_t *) buf = t;
        buf += 4;
    } while (--longs);
}
#endif
#endif

/*
 * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */
void MD5Init(struct MD5Context *ctx)
{
    /* Random seed */
    ctx->buf[0] = 0xf0321e32;
    ctx->buf[1] = 0xa450b330;
    /* Default seed */
    ctx->buf[2] = 0x98badcfe;
    ctx->buf[3] = 0x10325476;

    ctx->bits[0] = 0;
    ctx->bits[1] = 0;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
void MD5Update(struct MD5Context *ctx, unsigned char const *buf, unsigned len)
{
    uint32_t t;

    /* Update bitcount */

    t = ctx->bits[0];
    if ((ctx->bits[0] = t + ((uint32_t) len << 3)) < t)
        ctx->bits[1]++;         /* Carry from low to high */
    ctx->bits[1] += len >> 29;

    t = (t >> 3) & 0x3f;        /* Bytes already in shsInfo->data */

    /* Handle any leading odd-sized chunks */

    if (t) {
        unsigned char *p = (unsigned char *) ctx->in + t;

        t = 64 - t;
        if (len < t) {
            memcpy(p, buf, len);
            return;
        }
        memcpy(p, buf, t);
        byteReverse(ctx->in, 16);
        MD5Transform(ctx->buf, (uint32_t *) ctx->in);
        buf += t;
        len -= t;
    }
    /* Process data in 64-byte chunks */

    while (len >= 64) {
        memcpy(ctx->in, buf, 64);
        byteReverse(ctx->in, 16);
        MD5Transform(ctx->buf, (uint32_t *) ctx->in);
        buf += 64;
        len -= 64;
    }

    /* Handle any remaining bytes of data. */

    memcpy(ctx->in, buf, len);
}

/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern 
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
void MD5Final(unsigned char digest[16], struct MD5Context *ctx)
{
    unsigned count;
    unsigned char *p;

    /* Compute number of bytes mod 64 */
    count = (ctx->bits[0] >> 3) & 0x3F;

    /* Set the first char of padding to 0x80.  This is safe since there is
       always at least one byte free */
    p = ctx->in + count;
    *p++ = 0x80;

    /* Bytes of padding needed to make 64 bytes */
    count = 64 - 1 - count;

    /* Pad out to 56 mod 64 */
    if (count < 8) {
        /* Two lots of padding:  Pad the first block to 64 bytes */
        memset(p, 0, count);
        byteReverse(ctx->in, 16);
        MD5Transform(ctx->buf, (uint32_t *) ctx->in);

        /* Now fill the next block with 56 bytes */
        memset(ctx->in, 0, 56);
    } else {
        /* Pad block to 56 bytes */
        memset(p, 0, count - 8);
    }
    byteReverse(ctx->in, 14);

    /* Append length in bits and transform */
    ((uint32_t *) ctx->in)[14] = ctx->bits[0];
    ((uint32_t *) ctx->in)[15] = ctx->bits[1];

    MD5Transform(ctx->buf, (uint32_t *) ctx->in);
    byteReverse((unsigned char *) ctx->buf, 4);
    memcpy(digest, ctx->buf, 16);
    memset(ctx, 0, sizeof(ctx));        /* In case it's sensitive */
}

#ifndef ASM_MD5

/* The four core functions - F1 is optimized somewhat */

/* #define F1(x, y, z) (x & y | ~x & z) */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) F1(z, x, y)
#define F3(x, y, z) (x ^ y ^ z)
#define F4(x, y, z) (y ^ (x | ~z))

/* This is the central step in the MD5 algorithm. */
#ifdef __PUREC__
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f /*(x, y, z)*/ + data,  w = w<<s | w>>(32-s),  w += x )
#else
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f(x, y, z) + data,  w = w<<s | w>>(32-s),  w += x )
#endif

/*
 * The core of the MD5 algorithm, this alters an existing MD5 hash to
 * reflect the addition of 16 longwords of new data.  MD5Update blocks
 * the data and converts bytes into longwords for this routine.
 */
void MD5Transform(uint32_t buf[4], uint32_t const in[16])
{
    register uint32_t a, b, c, d;

    a = buf[0];
    b = buf[1];
    c = buf[2];
    d = buf[3];

#ifdef __PUREC__                /* PureC Weirdness... (GG) */
    MD5STEP(F1(b, c, d), a, b, c, d, in[0] + 0xd76aa478L, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[1] + 0xe8c7b756L, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[2] + 0x242070dbL, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[3] + 0xc1bdceeeL, 22);
    MD5STEP(F1(b, c, d), a, b, c, d, in[4] + 0xf57c0fafL, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[5] + 0x4787c62aL, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[6] + 0xa8304613L, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[7] + 0xfd469501L, 22);
    MD5STEP(F1(b, c, d), a, b, c, d, in[8] + 0x698098d8L, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[9] + 0x8b44f7afL, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[10] + 0xffff5bb1L, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[11] + 0x895cd7beL, 22);
    MD5STEP(F1(b, c, d), a, b, c, d, in[12] + 0x6b901122L, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[13] + 0xfd987193L, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[14] + 0xa679438eL, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[15] + 0x49b40821L, 22);

    MD5STEP(F2(b, c, d), a, b, c, d, in[1] + 0xf61e2562L, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[6] + 0xc040b340L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[11] + 0x265e5a51L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[0] + 0xe9b6c7aaL, 20);
    MD5STEP(F2(b, c, d), a, b, c, d, in[5] + 0xd62f105dL, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[10] + 0x02441453L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[15] + 0xd8a1e681L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[4] + 0xe7d3fbc8L, 20);
    MD5STEP(F2(b, c, d), a, b, c, d, in[9] + 0x21e1cde6L, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[14] + 0xc33707d6L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[3] + 0xf4d50d87L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[8] + 0x455a14edL, 20);
    MD5STEP(F2(b, c, d), a, b, c, d, in[13] + 0xa9e3e905L, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[2] + 0xfcefa3f8L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[7] + 0x676f02d9L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[12] + 0x8d2a4c8aL, 20);

    MD5STEP(F3(b, c, d), a, b, c, d, in[5] + 0xfffa3942L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[8] + 0x8771f681L, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[11] + 0x6d9d6122L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[14] + 0xfde5380cL, 23);
    MD5STEP(F3(b, c, d), a, b, c, d, in[1] + 0xa4beea44L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[4] + 0x4bdecfa9L, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[7] + 0xf6bb4b60L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[10] + 0xbebfbc70L, 23);
    MD5STEP(F3(b, c, d), a, b, c, d, in[13] + 0x289b7ec6L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[0] + 0xeaa127faL, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[3] + 0xd4ef3085L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[6] + 0x04881d05L, 23);
    MD5STEP(F3(b, c, d), a, b, c, d, in[9] + 0xd9d4d039L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[12] + 0xe6db99e5L, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[15] + 0x1fa27cf8L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[2] + 0xc4ac5665L, 23);

    MD5STEP(F4(b, c, d), a, b, c, d, in[0] + 0xf4292244L, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[7] + 0x432aff97L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[14] + 0xab9423a7L, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[5] + 0xfc93a039L, 21);
    MD5STEP(F4(b, c, d), a, b, c, d, in[12] + 0x655b59c3L, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[3] + 0x8f0ccc92L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[10] + 0xffeff47dL, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[1] + 0x85845dd1L, 21);
    MD5STEP(F4(b, c, d), a, b, c, d, in[8] + 0x6fa87e4fL, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[15] + 0xfe2ce6e0L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[6] + 0xa3014314L, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[13] + 0x4e0811a1L, 21);
    MD5STEP(F4(b, c, d), a, b, c, d, in[4] + 0xf7537e82L, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[11] + 0xbd3af235L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[2] + 0x2ad7d2bbL, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[9] + 0xeb86d391L, 21);
#else
    MD5STEP(F1, a, b, c, d, in[0] + 0xd76aa478, 7);
    MD5STEP(F1, d, a, b, c, in[1] + 0xe8c7b756, 12);
    MD5STEP(F1, c, d, a, b, in[2] + 0x242070db, 17);
    MD5STEP(F1, b, c, d, a, in[3] + 0xc1bdceee, 22);
    MD5STEP(F1, a, b, c, d, in[4] + 0xf57c0faf, 7);
    MD5STEP(F1, d, a, b, c, in[5] + 0x4787c62a, 12);
    MD5STEP(F1, c, d, a, b, in[6] + 0xa8304613, 17);
    MD5STEP(F1, b, c, d, a, in[7] + 0xfd469501, 22);
    MD5STEP(F1, a, b, c, d, in[8] + 0x698098d8, 7);
    MD5STEP(F1, d, a, b, c, in[9] + 0x8b44f7af, 12);
    MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1, 17);
    MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7be, 22);
    MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122, 7);
    MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193, 12);
    MD5STEP(F1, c, d, a, b, in[14] + 0xa679438e, 17);
    MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821, 22);

    MD5STEP(F2, a, b, c, d, in[1] + 0xf61e2562, 5);
    MD5STEP(F2, d, a, b, c, in[6] + 0xc040b340, 9);
    MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51, 14);
    MD5STEP(F2, b, c, d, a, in[0] + 0xe9b6c7aa, 20);
    MD5STEP(F2, a, b, c, d, in[5] + 0xd62f105d, 5);
    MD5STEP(F2, d, a, b, c, in[10] + 0x02441453, 9);
    MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681, 14);
    MD5STEP(F2, b, c, d, a, in[4] + 0xe7d3fbc8, 20);
    MD5STEP(F2, a, b, c, d, in[9] + 0x21e1cde6, 5);
    MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6, 9);
    MD5STEP(F2, c, d, a, b, in[3] + 0xf4d50d87, 14);
    MD5STEP(F2, b, c, d, a, in[8] + 0x455a14ed, 20);
    MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905, 5);
    MD5STEP(F2, d, a, b, c, in[2] + 0xfcefa3f8, 9);
    MD5STEP(F2, c, d, a, b, in[7] + 0x676f02d9, 14);
    MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8a, 20);

    MD5STEP(F3, a, b, c, d, in[5] + 0xfffa3942, 4);
    MD5STEP(F3, d, a, b, c, in[8] + 0x8771f681, 11);
    MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122, 16);
    MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380c, 23);
    MD5STEP(F3, a, b, c, d, in[1] + 0xa4beea44, 4);
    MD5STEP(F3, d, a, b, c, in[4] + 0x4bdecfa9, 11);
    MD5STEP(F3, c, d, a, b, in[7] + 0xf6bb4b60, 16);
    MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70, 23);
    MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6, 4);
    MD5STEP(F3, d, a, b, c, in[0] + 0xeaa127fa, 11);
    MD5STEP(F3, c, d, a, b, in[3] + 0xd4ef3085, 16);
    MD5STEP(F3, b, c, d, a, in[6] + 0x04881d05, 23);
    MD5STEP(F3, a, b, c, d, in[9] + 0xd9d4d039, 4);
    MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5, 11);
    MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8, 16);
    MD5STEP(F3, b, c, d, a, in[2] + 0xc4ac5665, 23);

    MD5STEP(F4, a, b, c, d, in[0] + 0xf4292244, 6);
    MD5STEP(F4, d, a, b, c, in[7] + 0x432aff97, 10);
    MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7, 15);
    MD5STEP(F4, b, c, d, a, in[5] + 0xfc93a039, 21);
    MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3, 6);
    MD5STEP(F4, d, a, b, c, in[3] + 0x8f0ccc92, 10);
    MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47d, 15);
    MD5STEP(F4, b, c, d, a, in[1] + 0x85845dd1, 21);
    MD5STEP(F4, a, b, c, d, in[8] + 0x6fa87e4f, 6);
    MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0, 10);
    MD5STEP(F4, c, d, a, b, in[6] + 0xa3014314, 15);
    MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1, 21);
    MD5STEP(F4, a, b, c, d, in[4] + 0xf7537e82, 6);
    MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235, 10);
    MD5STEP(F4, c, d, a, b, in[2] + 0x2ad7d2bb, 15);
    MD5STEP(F4, b, c, d, a, in[9] + 0xeb86d391, 21);
#endif

    buf[0] += a;
    buf[1] += b;
    buf[2] += c;
    buf[3] += d;
}

#endif
